### PR TITLE
Waveguide crossing inverse design example

### PR DESCRIPTION
**Pull request by smartalecH**
Here we add a python inverse-design example that maximizes the transmission of a waveguide crossing. In particular, it's meant to showcase the new (first-order accurate) smoothed projection functionality. Using this example, you can:

**Perform shape optimization:**

image

**Perform topology optimization:**

image

**Analyze the norm of the gradient as $\beta \to \infty$ for both the smoothed case and non-smoothed case:**

image

**Analyze the convergence properties of a shape optimization problem at a finite beta ($\beta=64$ here) for both the smoothed case and non-smoothed case:**

image

### CODE REVIEW

**Review by oskooi — CHANGES_REQUESTED** [review]
Is it really necessary to have the two functions `run_shape_optimization` and `run_topology_optimization` be separate? They share a lot of code and can probably be combined into a single function.

**Comment by smartalecH** [comment]
Here is an example that runs for 1000 iterations. Interesting how the non-smoothing case eventually breaks through the floor and keeps descending.

image

File: python/examples/waveguide_crossing.py [hunk 1]
```diff
@@ -0,0 +1,615 @@
...
+        ),
+    ]
+
+    def J(input, output):
+        """Simple objective function to minimize loss."""
+        return 1 - npa.power(npa.abs(output / input), 2)
+
+    opt = mpa.OptimizationProblem(
+        simulation=sim,
+        maximum_run_time=500,
```
[the remark this answers is not in the document]
Comment (smartalecH) [reply python/examples/waveguide_crossing.py:175]: Ya I actually had the same idea at first and wrote my first draft with them combined. But there were too many flags to pass and case logic needed, such that it was really hard to parse.

So from a pedagogical standpoint (these are tutorials, afterall) I think this is better. The amount of "duplicate" code is actually really minimal anyways. Just setting up the optimization problem (which is already factored out as it's on function).

File: python/examples/waveguide_crossing.py [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+        results: Structure to store the simulated FOM each iteration.
+
+    Returns:
+        The FOM value at the current iteration.
+    """
+    grid_size = opt.design_regions[0].design_parameters.grid_size
+    Nx, Ny = int(grid_size.x), int(grid_size.y)
+    data.append(mapping(x).copy().reshape(Nx, Ny))
+
+    f0, dJ_du = opt([mapping(x)])
```
[the remark this answers is not in the document]
Comment (smartalecH) [reply python/examples/waveguide_crossing.py:241]: I don't see any variables with camel case here? in fact they already align with google's convention (`grid_size`), right?

sure we can change `f0` and `dJ_du` but let's do that in a separate PR to _all_ the examples. Don't want to be mixing conventions midway through, right?

File: python/examples/waveguide_crossing.py [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+    plt.legend()
+    plt.xlabel("Optimization iteration")
+    plt.ylabel("FOM")
+    plt.show()
+
+    return results, results_smoothed
+
+
+if __name__ == "__main__":
+    run_shape_optimization(resolution=25.0, beta=np.inf, maxeval=30)
```
[the remark this answers is not in the document]
Comment (smartalecH) [reply python/examples/waveguide_crossing.py:605]: Not used here so NA

File: python/adjoint/filters.py [hunk 4]
```diff
@@ -694,9 +695,164 @@ def tanh_projection(x: np.ndarray, beta: float, eta: float) -> np.ndarray:
...
+        # Note that backpropagating through here can produce NaNs. So we
+        # manually specify the step function to keep the gradient clean.
+        return npa.where(x > eta, 1.0, 0.0)
+    else:
+        return (npa.tanh(beta * eta) + npa.tanh(beta * (x - eta))) / (
+            npa.tanh(beta * eta) + npa.tanh(beta * (1 - eta))
+        )
+
+
+def smoothed_projection(
```
[the remark this answers is not in the document]
Comment (smartalecH) [reply python/adjoint/filters.py:708]: rebased, thanks

File: python/examples/waveguide_crossing.py [hunk 5]
```diff
@@ -0,0 +1,615 @@
...
+# Importantly, this particular example highlights some of the ways one can use
+# the novel smoothed projection function to perform both shape and topology
+# optimization.
+
+
+from typing import Callable, List, Optional, Tuple
+
+import meep.adjoint as mpa
+import nlopt
+import numpy as np
```
[the remark this answers is not in the document]
Comment (smartalecH) [reply python/examples/waveguide_crossing.py:16]: Right but autograd's numpy is missing a lot of functionality. It's not like jax's which is close to feature complete. There are some functions which require "classic numpy."

Plus I actually think it's good to use classic numpy for code you don't want to run through an AD, that way you can throw/catch errors.

**Comment by mochen4** [comment]
As discussed, if it is a greyscale issue, you can try using the ``damping`` term in the MaterialGrid, and it really helps with binarization. From my experience, typically ``damping=0.05ω`` is sufficient.

**Comment by smartalecH** [comment]
If we look at the final designs for both cases, we see that the optimizer is _not_ simply relying on making the non-smoothed case less binary. Rather, it continues to perturb the geometry.

image

**Comment by stevengj** [comment]
I wonder if it's getting stuck from the subpixel-smoothed structure being only piecewise differentiable?

**Comment by smartalecH** [comment]
So here is a plot of the gradients during the "flat region" of the smoothing optimization:

image

What's interesting, is that the gradients for the smoothed case are _antisymmetric_ across _x_ and _y_... this is problematic though because our mapping function forces a symmetry across _x_ and _y_... so the optimizer can't actually achieve the change it wants too (i.e. there may be a bug somewhere breaking down the gradient accuracy?)

**Comment by stevengj** [comment]
The alternating signs of the gradient would cause the changes to cancel out to first order, so if that's due to a sign error it would be a problem.

**Comment by smartalecH** [comment]
@stevengj if I monitor the output of CCSA during the optimization, `rho` indeed blows up.

During the stalled region, it's not necessarily stuck in the same outer iteration. But `rho` keeps climbing (eg 1e15) and it does take more inner iterations to converge than the rest of the optimization.

**Comment by stevengj** [comment]
Some things to try:

1. Check smoothness numerically.  Define ρ to be e.g. a Gaussian centered at (x,0), which projects to a cylinder, and plot the smoothed/projected ρ̂  at some pixel as a function of x — it should go smoothly (or at least with a continuous 1st derivative) from 0 to 1 as the cylinder passes through the pixel location.

2. Replace the smoothing function
image
with some function of our choosing, e.g. a Gaussian or a twice-differentiable "bump" function with compact support (on a sphere).    Implicitly, this defines a smoothing kernel that you could reverse-engineer if you wanted (since the smoothing kernel integrated over the spherical cap gives F_c, you should be able to take some derivatives of the latter to get the former)

File: python/adjoint/filters.py [hunk 6]
```diff
@@ -694,9 +695,164 @@ def tanh_projection(x: np.ndarray, beta: float, eta: float) -> np.ndarray:
...
+    x_plus_eff = npa.where(
+        (d > 0),
+        x_plus,
+        x_plus_eff_pert,
+    )
+
+    # Only apply smoothing to interfaces
+    x_projected_smoothed = (1 - fill_factor_eff) * x_minus_eff + (
+        fill_factor_eff
+    ) * x_plus_eff
```
Comment (smartalecH) [inline python/adjoint/filters.py:850]: > Check smoothness numerically.

Hmm so after running @stevengj's experiment, we can clearly see the discontinuity at play here:

image

To be precise, we still get the continuous "sigmoid-like" transition (from the 2d case of $F_c$), but it's not being scaled properly. So it can't transition properly from/to 0/1.

The code above shows how this scaling is performed.

Here's what happens if we use `beta=8`:

image

Comment (smartalecH) [reply python/adjoint/filters.py]: Ah Ok I think I've figured it out. Essentially, I was forgetting to _project_ our endpoints too (which affects the scaling). When I do this, We get the following results:

`beta=np.inf`
image

`beta=8`
image

(note that the derivative is still not continuous in these cases. But that's somewhat expected, since we don't take the derivative at the edges into account when constructing $F_c$. So a more "intelligent" choice for $F_c$ here might be something that looks more like a Cubic hermite spline or something.)

The gradient convergence still looks good:
image


And the convergence plot no longer stalls:
Figure_1

Comment (stevengj) [reply python/adjoint/filters.py]: Will this affect https://github.com/NanoComp/photonics-opt-testbed/pull/63, since that used a Julia port of this code?

Comment (smartalecH) [reply python/adjoint/filters.py]: > since that used a Julia port of this code?

Yep I'm talking with @hammy4815 offline and he's trying out an updated version to see if this resolves some of the challenges he was seeing.
